src/search/counter_moves.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "movegen/move_types.h"
+#include <algorithm>
+
+namespace search {
+
+class CounterMoves {
+public:
+    void update(const movegen::Move& prevMove, const movegen::Move& counterMove)
+    {
+        /* only quiet moves */
+        if (counterMove.isCapture()) {
+            return; // nothing to do
+        }
+        m_counterMoves.at(prevMove.fromPos()).at(prevMove.toPos()) = counterMove;
+    }
+
+    movegen::Move get(const movegen::Move& prevMove) const
+    {
+        return m_counterMoves.at(prevMove.fromPos()).at(prevMove.toPos());
+    }
+
+    void reset()
+    {
+        for (auto& row : m_counterMoves)
+            std::ranges::fill(row, movegen::nullMove());
```

Comment:
if you add a type eg: `using CounterEntry = std::array<movegen::Move, s_amountSquares>` you can prettify `m_counterMoves` type and avoid iterating here:
```suggestion
        std::ranges::fill(m_counterMoves, CounterEntry {});
```